main/src/ca/uwaterloo/flix/language/phase/Typer.scala

```diff
@@ -305,270 +338,279 @@ object Typer extends Phase[ResolvedAst.Root, TypedAst.Root] {
   /**
     * Infers the type of the given expression `exp0`.
     */
-  private def inferExp(exp0: ResolvedAst.Expression, root: ResolvedAst.Root)(implicit flix: Flix): InferMonad[(Type, Type)] = {
+  private def inferExp(exp0: ResolvedAst.Expression, root: ResolvedAst.Root)(implicit flix: Flix): InferMonad[(Set[TypedAst.TypeConstraint], Type, Type)] = {
 
     /**
       * Infers the type of the given expression `exp0` inside the inference monad.
       */
-    def visitExp(e0: ResolvedAst.Expression): InferMonad[(Type, Type)] = e0 match {
+    def visitExp(e0: ResolvedAst.Expression): InferMonad[(Set[TypedAst.TypeConstraint], Type, Type)] = e0 match {
```

Comment:
I agree with this signature extension. 

My only concern is whether it should be a set. It is *effectively* a set, but in the literature it is often described as a list (with a rule to eliminate duplicates).

Moreover, we try to avoid sets in the compiler because they are unordered. It could be a sorted set (then iteration is deterministic), but then you need an ordering on types (ugh).

I would suggest that we replace it by a list.

Reply:
We could use a ListSet for built-in deduping. Or would it be better to eliminate all duplicates at once in a later step instead of checking each insertion?

Reply:
I think elimination of duplicate constraints should be explicit (somewhere).

Reply:
Also, I believe we now have types that are equal, but not syntactically equal. 

For example (if stupid):

```
a -> b & F
a -> b & (not e and e)
```

so we need to some more heavy lifting somewhere. Better to make it explicit then.